File: go/merkle_test.go
package ledgerrag

import (
	"bytes"
	"testing"
)

func TestSingleLeafRootIsLeafHash(t *testing.T) {
	tree, err := NewMerkleTree([][]byte{[]byte("only")})
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(tree.Root(), HashLeaf([]byte("only"))) {
		t.Fatal("single-leaf root must equal the leaf hash")
	}
}

func TestHashNodesOrderMatters(t *testing.T) {
	a := HashLeaf([]byte("a"))
	b := HashLeaf([]byte("b"))
	if bytes.Equal(HashNodes(a, b), HashNodes(b, a)) {
		t.Fatal("node hash must depend on child order")
	}
}

func TestLeafAndNodeHashesAreDomainSeparated(t *testing.T) {
	// A leaf hash must never collide with a node hash of the same bytes.
	if bytes.Equal(HashLeaf([]byte("x")), HashNodes([]byte("x"), nil)) {
		t.Fatal("leaf and node hashing must be domain-separated")
	}
}

func TestValidInclusionProofVerifiesAllIndices(t *testing.T) {
	leaves := [][]byte{}
	for i := 0; i < 7; i++ { // odd count on purpose
		leaves = append(leaves, []byte{byte('a' + i)})
	}
	tree, err := NewMerkleTree(leaves)
	if err != nil {
		t.Fatal(err)
	}
	for i, leaf := range leaves {
		proof, err := tree.InclusionProof(i)
		if err != nil {
			t.Fatal(err)
		}
		if !VerifyInclusion(leaf, proof, tree.Root()) {
			t.Fatalf("leaf %d should verify", i)
		}
	}
}

func TestTamperedLeafIsRejected(t *testing.T) {
	leaves := [][]byte{[]byte("chunk-0"), []byte("chunk-1"), []byte("chunk-2"), []byte("chunk-3"), []byte("chunk-4")}
	tree, err := NewMerkleTree(leaves)
	if err != nil {
		t.Fatal(err)
	}
	proof, err := tree.InclusionProof(2)
	if err != nil {
		t.Fatal(err)
	}
	if VerifyInclusion([]byte("chunk-2-TAMPERED"), proof, tree.Root()) {
		t.Fatal("tampered leaf must not verify")
	}
}

func TestWrongRootIsRejected(t *testing.T) {
	tree, _ := NewMerkleTree([][]byte{[]byte("a"), []byte("b"), []byte("c"), []byte("d")})
	other, _ := NewMerkleTree([][]byte{[]byte("a"), []byte("b"), []byte("c"), []byte("e")})
	proof, err := tree.InclusionProof(3)
	if err != nil {
		t.Fatal(err)
	}
	if VerifyInclusion([]byte("d"), proof, other.Root()) {
		t.Fatal("proof must not verify against a different root")
	}
}

func TestProofLengthIsLogarithmic(t *testing.T) {
	leaves := [][]byte{}
	for i := 0; i < 8; i++ {
		leaves = append(leaves, []byte{byte('0' + i)})
	}
	tree, _ := NewMerkleTree(leaves)
	proof, err := tree.InclusionProof(0)
	if err != nil {
		t.Fatal(err)
	}
	if len(proof) != 3 { // log2(8)
		t.Fatalf("expected proof length 3, got %d", len(proof))
	}
}

func TestEmptyTreeErrors(t *testing.T) {
	if _, err := NewMerkleTree(nil); err == nil {
		t.Fatal("empty tree must error")
	}
}

func TestInclusionProofOutOfRangeErrors(t *testing.T) {
	tree, _ := NewMerkleTree([][]byte{[]byte("a"), []byte("b")})
	if _, err := tree.InclusionProof(-1); err == nil {
		t.Fatal("negative index must error")
	}
	if _, err := tree.InclusionProof(2); err == nil {
		t.Fatal("index >= leaf count must error")
	}
}

func TestOddLeafPromotionLastLeaf(t *testing.T) {
	// A 3-leaf tree pairs the last leaf with itself when promoting.
	leaves := [][]byte{[]byte("a"), []byte("b"), []byte("c")}
	tree, _ := NewMerkleTree(leaves)
	proof, err := tree.InclusionProof(2)
	if err != nil {
		t.Fatal(err)
	}
	if !VerifyInclusion([]byte("c"), proof, tree.Root()) {
		t.Fatal("promoted last leaf should still verify")
	}
}

func TestCorruptingProofSiblingIsRejected(t *testing.T) {
	leaves := [][]byte{[]byte("a"), []byte("b"), []byte("c"), []byte("d")}
	tree, _ := NewMerkleTree(leaves)
	proof, err := tree.InclusionProof(1)
	if err != nil {
		t.Fatal(err)
	}
	corrupted := make([]ProofStep, len(proof))
	copy(corrupted, proof)
	corrupted[0].Sibling = append([]byte{}, corrupted[0].Sibling...)
	corrupted[0].Sibling[0] ^= 0x01
	if VerifyInclusion([]byte("b"), corrupted, tree.Root()) {
		t.Fatal("corrupted sibling must not verify")
	}
}

func TestManualTwoLeafRoot(t *testing.T) {
	a := []byte("a")
	b := []byte("b")
	root := HashNodes(HashLeaf(a), HashLeaf(b))
	proof := []ProofStep{{Sibling: HashLeaf(b), Side: "right"}}
	if !VerifyInclusion(a, proof, root) {
		t.Fatal("hand-built proof should verify")
	}
	if VerifyInclusion([]byte("TAMPERED"), proof, root) {
		t.Fatal("wrong leaf must not verify against hand-built root")
	}
}

File: ts/src/merkle.test.ts
import { describe, expect, it } from "vitest";

import { hashLeaf, hashNodes, MerkleTree, verifyInclusion, type ProofStep } from "./merkle.ts";

function leaves(items: string[]): Buffer[] {
  return items.map((s) => Buffer.from(s, "utf8"));
}

describe("merkle", () => {
  it("root of a single leaf equals its leaf hash", () => {
    const tree = new MerkleTree(leaves(["only"]));
    expect(tree.root().equals(hashLeaf(Buffer.from("only")))).toBe(true);
  });

  it("hashNodes is order sensitive", () => {
    const a = hashLeaf(Buffer.from("a"));
    const b = hashLeaf(Buffer.from("b"));
    expect(hashNodes(a, b).equals(hashNodes(b, a))).toBe(false);
  });

  it("leaf and node hashes are domain separated", () => {
    expect(hashLeaf(Buffer.from("x")).equals(hashNodes(Buffer.from("x"), Buffer.alloc(0)))).toBe(
      false,
    );
  });

  it("valid inclusion proofs verify for every index (odd leaf count)", () => {
    const items = leaves(["a", "b", "c", "d", "e", "f", "g"]);
    const tree = new MerkleTree(items);
    items.forEach((leaf, i) => {
      expect(verifyInclusion(leaf, tree.inclusionProof(i), tree.root())).toBe(true);
    });
  });

  it("a tampered leaf is rejected", () => {
    const tree = new MerkleTree(leaves(["chunk-0", "chunk-1", "chunk-2", "chunk-3", "chunk-4"]));
    const proof = tree.inclusionProof(2);
    expect(verifyInclusion(Buffer.from("chunk-2-TAMPERED"), proof, tree.root())).toBe(false);
  });

  it("a proof does not verify against a different root", () => {
    const tree = new MerkleTree(leaves(["a", "b", "c", "d"]));
    const other = new MerkleTree(leaves(["a", "b", "c", "e"]));
    const proof = tree.inclusionProof(3);
    expect(verifyInclusion(Buffer.from("d"), proof, other.root())).toBe(false);
  });

  it("proof length is logarithmic in the leaf count", () => {
    const tree = new MerkleTree(leaves(["0", "1", "2", "3", "4", "5", "6", "7"]));
    expect(tree.inclusionProof(0).length).toBe(3);
  });

  it("an empty tree throws", () => {
    expect(() => new MerkleTree([])).toThrow();
  });

  it("an out-of-range inclusion proof throws", () => {
    const tree = new MerkleTree(leaves(["a", "b"]));
    expect(() => tree.inclusionProof(2)).toThrow();
  });

  it("the promoted last leaf of an odd level still verifies", () => {
    const tree = new MerkleTree(leaves(["a", "b", "c"]));
    expect(verifyInclusion(Buffer.from("c"), tree.inclusionProof(2), tree.root())).toBe(true);
  });

  it("corrupting a proof sibling is rejected", () => {
    const tree = new MerkleTree(leaves(["a", "b", "c", "d"]));
    const proof = tree.inclusionProof(1);
    proof[0].sibling[0] ^= 0x01;
    expect(verifyInclusion(Buffer.from("b"), proof, tree.root())).toBe(false);
  });

  it("verifies against a hand-built two-leaf root", () => {
    const root = hashNodes(hashLeaf(Buffer.from("a")), hashLeaf(Buffer.from("b")));
    const proof: ProofStep[] = [{ sibling: hashLeaf(Buffer.from("b")), side: "right" }];
    expect(verifyInclusion(Buffer.from("a"), proof, root)).toBe(true);
    expect(verifyInclusion(Buffer.from("TAMPERED"), proof, root)).toBe(false);
  });
});
